inspirehep/modules/multieditor/actions.py

```diff
@@ -0,0 +1,308 @@
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, print_function, division
+import json
+import os
+import re
+from collections import namedtuple, deque
+from inspire_schemas.api import load_schema
+
+
+Action = namedtuple('Action', 'keys, selected_action, value,'
+                              ' update_regex, values_to_check, where_regex, where_keys, where_value')
+
+
+class Addition(object):
+    def __init__(self, keys, value, update_regex, values_to_check,
+                 where_regex, where_keys, where_values):
+        self.keys = keys
+        self.value = value
+        self.update_regex = update_regex
+        self.values_to_check = values_to_check
+        self.where_regex = where_regex
+        self.where_keys = where_keys
+        self.where_values = where_values
+
+    def apply_action(self, record, schema, position = 0, checked=False):
+        """Recursive function to change a record object."""
+        key = self.keys[position]
+        where_key = None
+        if len(self.where_keys) > position:
+            where_key = self.where_keys[position]
+        new_schema = {}
+        if schema:  # for testing purposes
+            if schema['type'] == 'object':
+                new_schema = schema['properties'][key]
+            elif schema['type'] == 'array':
+                new_schema = schema['items']['properties'][key]
+
+        if where_key:
+            if not checked and key != where_key: # fixme possible refactor
+                if not check_value(record=record, keys=self.where_keys, values_to_check=self.where_values,
+                                   regex=self.where_regex, position=position):
+                    return
+                checked = True
+        if not record.get(key): # adding to object as well
+            creation_keys = self.keys[position:]
+            record.update(create_schema_record(schema, creation_keys, self.value))
+            return
+        if position + 1 == len(self.keys):
+            if isinstance(record[key], list):
+                record[key].append(self.value)
+        else:
+            if isinstance(record[key], list):
+                for array_record in record[key]:
+                    self.apply_action(array_record, new_schema, position+1, checked)
+            else:
+                self.apply_action(record[key], new_schema, position+1, checked)
```

Comment:
This is not a good class, because it's just a function call in disguise. Watch this talk to get what I mean: https://www.youtube.com/watch?v=o9pEzgHorH0.